pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/ConsistentDataPushUtils.java

```diff
@@ -0,0 +1,296 @@
+      String segmentName = fileName.substring(0, fileName.length() - Constants.TAR_GZ_FILE_EXT.length());
+      segmentNames.add(segmentName);
+    }
+    return segmentNames;
+  }
+
+  public static TableConfig getTableConfig(SegmentGenerationJobSpec spec) {
+    String rawTableName = spec.getTableSpec().getTableName();
+    TableConfig tableConfig =
+        SegmentGenerationUtils.getTableConfig(spec.getTableSpec().getTableConfigURI(), spec.getAuthToken());
+    Preconditions.checkState(tableConfig != null, "Failed to find table config for table: %s", rawTableName);
+    return tableConfig;
+  }
+
+  public static boolean consistentDataPushEnabled(SegmentGenerationJobSpec spec) {
+    TableConfig tableConfig = getTableConfig(spec);
+    // Enable consistent data push only if "consistentDataPush" is set to true in batch ingestion config and the
+    // table is REFRESH use case.
+    return "REFRESH".equalsIgnoreCase(IngestionConfigUtils.getBatchSegmentIngestionType(tableConfig))
+        && IngestionConfigUtils.getBatchSegmentIngestionConsistentDataPushEnabled(tableConfig);
+  }
+
+  public static boolean uniqueSegmentNameDisabled(SegmentGenerationJobSpec spec) {
+    TableConfig tableConfig = getTableConfig(spec);
+    return IngestionConfigUtils.getBatchSegmentIngestionDisableUniqueSegmentNames(tableConfig);
+  }
+
+  /**
+   * Returns a map of controller URI to a list of existing OFFLINE segments.
+   */
+  public static Map<URI, List<String>> getSelectedOfflineSegments(SegmentGenerationJobSpec spec, String rawTableName) {
+    Map<URI, List<String>> uriToOfflineSegments = new HashMap<>();
+    for (PinotClusterSpec pinotClusterSpec : spec.getPinotClusterSpecs()) {
+      URI controllerURI;
+      List<String> offlineSegments;
+      try {
+        controllerURI = new URI(pinotClusterSpec.getControllerURI());
+        offlineSegments =
+            FILE_UPLOAD_DOWNLOAD_CLIENT.getSegments(controllerURI, rawTableName, TableType.OFFLINE.toString(), true);
+        uriToOfflineSegments.put(controllerURI, offlineSegments);
+      } catch (URISyntaxException e) {
+        throw new RuntimeException("Got invalid controller uri - '" + pinotClusterSpec.getControllerURI() + "'");
+      } catch (HttpErrorStatusException | IOException e) {
+        e.printStackTrace();
+      }
+    }
+    return uriToOfflineSegments;
+  }
+
+  public static void configureSegmentPostfix(SegmentGenerationJobSpec spec) {
+    SegmentNameGeneratorSpec segmentNameGeneratorSpec = spec.getSegmentNameGeneratorSpec();
+    if (consistentDataPushEnabled(spec) && !uniqueSegmentNameDisabled(spec)) {
+      // Add the current timestamp as the segment postfix to make segment name unique.
+      if (segmentNameGeneratorSpec == null) {
+        segmentNameGeneratorSpec = new SegmentNameGeneratorSpec();
+      }
+      segmentNameGeneratorSpec.addConfig(SEGMENT_NAME_POSTFIX, Long.toString(System.currentTimeMillis()));
```

Comment:
Users may configure their own postfix? Better to append the timestamp to the end if there's an existing postfix?

Reply:
Addressed.